initial cleanup (#8)
informalsystems/cross-chain-validation: spec/valset-update-protocol.md

  - Baby blockchain: Baby blockchain is a blockchain that is being secured by the parent blockchain. In other words, validators that secure and operate the baby blockchain are bonded on the parent blockchain. Any misbehavior of a validator at the baby blockchain is punished at the parent blockchain (i.e., the validator is slashed at the parent blockchain).
  - IBC communication: IBC communication allows the parent and baby blockchain to communicate. We assume that the IBC communication is ordered and provides "timeout" and "acknowledgements" mechanisms.

### Properties

This subsection is devoted to defining properties that the Cross-Chain Validation protocol ensures.
Recall that the parent blockchain has an ability to demand a change to the validator set of the baby chain.
Moreover, we want to ensure the stake of validators of the baby blockchain are "frozen" at the parent chain.

> Hence, we aim to achieve the following properties:

Reviewer: I think we need a safety property: If the baby chain sets its validator set from V1 to V2, then the staking module of the parent chain requested the change to V2 after V1.

I guess, if we don't assume the Vs to be unique, we need to incorporate the sequence numbers into the safety and liveness properties.

Author: Yes, agreed!

> - *Liveness*: If the parent blockchain demands a change to the validator set of the baby chain to some set *V* of validators and the IBC communication successfully relays this demand to the baby blockchain, then the validator set of the baby blockchain is eventually set to *V* or to some set *V'*, such that the parent blockchain demanded a change to the validator set of the baby chain to *V'* **after** the demand for *V*.

Reviewer: ```suggestion
- *Liveness*: If the staking module of the parent blockchain demands a change to the validator set of the baby chain to some set *V* of validators and the IBC communication successfully relays this demand to the baby blockchain, then the validator set of the baby blockchain is eventually set to *V* or to some set *V'*, such that the parent blockchain demanded a change to the validator set of the baby chain to *V'* **after** the demand for *V*.
```

- *Stake safety*: If a validator *v* belongs to a validator set of the baby blockchain, then the stake of *v* is frozen at the parent blockchain. Moreover, if the stake of *v* is unfrozen at the parent blockchain, the unbonding period has elapsed for *v* at the baby blockchain.

> - *Stake liveness*: If the stake of a validator *v* is bonded at the parent blockchain, eventually the stake will be unbonded (some stake might be slashed).

Reviewer: I guess we need to add in the precondition that the validator set change is requested by the staking module. If such a request never happens, why should the stake be unbonded?

Author: Yes, good point! 
If we want to be completely precise, we need to take time into account.
In other words, we should add something like this: "the validator set change that does not contain v takes place at the baby blockchain and sufficiently long time elapses before another change validator set that contains v is issued by the staking logic of the parent blockchain".
This "sufficiently long time" is simply there to ensure that there is a time window for the stake to indeed be unbonded.

Author: Shall we complicate the property this much or is your suggestion enough for the spec?

Reviewer: I guess your property is about stability. If at the parent chain the VS changes stop for x time, then... I guess we can add it in addition to the safety. We can discuss tomorrow.

### Closer look at the IBC communication

> An IBC connection assumes two parties (the respective blockchains) involved in the communication. However, it also assumes a relayer which handles message transmissions between the two blockchains. The relayer carries a central responsibility in ensuring communication between the two parties.

Reviewer: I guess at the application level we consider "IBC channels" rather than "IBC connections"?

Author: I guess so. To be honest, I still do not consider myself an IBC expert, so please intervene whenever my writing/understanding of the IBC is wrong.

  - User-driven state transitions: These transitions "start" the entire process of changing the validator set of the baby blockchain.
  We assume that the staking module expresses the will to change the validator set of the baby blockchain. It is done at the End-Block method (as in the "single blockchain" scenario).

  - Transaction-driven state transitions: These are the state transitions that are driven by the relayer. Namely, these transitions are activated since communication between the two blockchains takes place. E.g., some packet is received via the IBC communication, a timeout has elapsed for a packet, acknowledgment for a packet is received.

  - Time-driven state transitions: These transitions are activated since some time has elapsed. As we will present in the rest of the document, time-driven state transitions help us determine when the unbonding period, measured at the baby blockchain, has elapsed for a validator.

  In the rest of the section, we will discuss the aforementioned state transitions in more detail.

## State Machine for a Single Validator Set Change demand

> When we say that there is a *single* validator set change demand, we assume that this demand will be a single committed validator set change demand in a block at the baby blockchain.
> If this is indeed the case, the validator set change indeed takes place at the baby blockchain.
> In other words, the validator set of the baby blockchain changes to the validator set from the demand.

Reviewer: ```suggestion
For simplicity in presentation, in this section we consider a *single* validator set change demand, that is, we assume that this demand will be a single committed validator set change demand in a block at the parent blockchain.
(If the relayer successfully transmits the protocol packet) this demand will result the validator set change indeed taking place at the baby blockchain.
However, in principle multiple demands can be issued before the first of those is finished. This leads to concurrency, and intermediate validator set changes are not visible in the validator sets of the baby chain. Note, that this does not influence the (un)staking logic. We will discuss concurrency effects below.
```

We now present the state machine for this case.
![image](../images/data_flow.png)

> 

Reviewer: It would be good to describe the flow in English. It need not be very detailed, but the general working (of at least the lucky path) should be clear: The staking module issues a request -> a packet is send -> a packet is received, which starts a timeout, etc.

Author: Sure, the brief description is added.

## Function Definitions

### Parent blockchain

This subsection will present the functions executed at the parent blockchain.

```golang
// invoked by the staking module of the parent blockchain to
// express will to modify the validator set of the baby blockchain;
// executed in the End-Block method; similarly to the "normal, single-blockchain" case
func changeValidatorSet(
  babyChainId: ChainId
  valSet: Validator[]) {
  // freeze stake of validators from valSet associated with the seqNum; seqNum is a variable of the staking module
  for each (v in valSet) {
      freezeStake(v, seqNum)

>   }  

Author: Done.

>   }  

Author: Just one comment here: If we want to give only the updates as the input, the parent blockchain must know the validator set of the baby blockchain. Then, we will need to introduce an additional packet sent from the baby to the parent to signalize that the new validator set is indeed operating at the baby blockchain. Do we want to do that?

>   }  

Reviewer: I am not sure I understand the comment. Let's discuss tomorrow.

>   }  

Author: I guess I overthought this, this additional packet is not really needed. But yes, we can talk tomorrow. Thanks.

  // create the ChangeValidatorSet packet
  ChangeValidatorSet data = ChangeValidatorSet{valSet, seqNum++}

  // obtain the destination port of the baby blockchain
  destPort = getPort(babyChainId)

  // send the packet
  handler.sendPacket(Packet{timeoutHeight, timeoutTimestamp, destPort, destChannel, sourcePort, sourceChannel, data}, getCapability("port"))
}
```

- Expected precondition
  - There exists a blockchain with *babyChainId* identifier
  - All validators from *valSet* are validators at the parent blockchain
- Expected postcondition
  - Stake of each validator from *valSet* is frozen and associated with this demand (via *seqNum*)
  - The packet containing information about this change validator set demand is created
- Error condition
  - If the precondition is violated

```golang
// executed at the parent blockchain to handle a delivery of the IBC packet
func onRecvPacket(packet: Packet) {
  // the packet is of UnbondingOver type

>   assert(Packet = UnbondingOver)

Reviewer: shouldn't that be "Packet.type = "?

Author: Yes, absolutely! Fixed.

  valSet = packet.valSet
  seqNum = packet.seqNum

  // unfreeze stake of validators associated with this seqNum
  for each (v in valSet) {
    unfreezeStake(v, seqNum)

>   }

Author: Yes, done!

  // construct the default acknowledgment
  ack = defaultAck(UnbondingOver)
  return ack
}
```

- Expected precondition
  - The *ChangeValidatorSet* packet is sent to the baby blockchain before this packet is received
  - The packet is of the *UnbondingOver* type
- Expected postcondition
  - Stake of each validator from *valSet* associated with a sequence number *seqNum* is unfrozen
  - The default acknowledgment is created
- Error condition
  - If the precondition is violated

```golang
// called once a sent packet has timed-out
function onTimeoutPacket(packet: Packet) {
  // the packet is of ChangeValidatorSet type
  assert(packet = ChangeValidatorSet)

  valSet = packet.valSet
  seqNum = packet.seqNum

  // unfreeze stake of validators associated with this seqNum
  for each (v in valSet) {
    unfreezeStake(v, seqNum)
  }

> }

Author: Done.

```

- Expected precondition
  - The *packet* has timed out
- Expected postcondition
  - Stake of each validator from *valSet* associated with a sequence number *seqNum* is unfrozen
- Error condition
  - If the precondition is violated  

### Baby blockchain

```golang
// executed at the baby blockchain to handle a delivery of the IBC packet
func onRecvPacket(packet: Packet) {
  // the packet is of ChangeValidatorSet type
  assert(Packet = ChangeValidatorSet)

  valSet = packet.valSet
  seqNum = packet.seqNum

  // inform the staking module of the new validator set change demand

>   stakingModule.queueValidatorSetChange(valSet, seqNum)

Author: Please, see the comment about updates above.

- Expected postcondition
  - The staking module is queues the new validator set change demand
  - The default acknowledgment is created
- Error condition
  - If the precondition is violated

```golang
// End-Block method executed at the end of each block
func endBlock(block: Block) {
  // get time
  time = block.time

  // get the old validator set
  oldValSet, oldSeqNum = block.validatorSet

  // start unbonding for the old validator set and all validators set specified in the current block, except for last
  stakingModule.startUnbonding(oldValSet, oldSeqNum, time)
  while (stakingModule.sizeValidatorSetChangeQueue() > 1) {
    stakingModule.startUnbonding(stakingModule.dequeueValidatorSetChange(), time)
  }

  // finish unbonding for mature validator sets
  for each (valSet, seqNum in stakingModule.finishUnbonding(time)) {
    // create the UnbondingOver packet
    UnbondingOver data = UnbondingOver{valSet, seqNum}

    // obtain the destination port of the baby blockchain
    destPort = getPort(parentChainId)

    // send the packet
    handler.sendPacket(Packet{timeoutHeight, timeoutTimestamp, destPort, destChannel, sourcePort, sourceChannel, data}, getCapability("port"))

>   }

Author: Done!

- ChangeValidatorSet(validatorSet, seqNum): Packet sent by the parent blockchain to the baby blockchain to express the demand of the parent blockchain to modify the validator set of the baby blockchain.
Validator set modifications are defined in the *valSetUpdate* parameter, whereas the *seqNum* represents the unique identifier of the demand.

  *Remark:* There exists the acknowledgment packet for the ChangeValidatorSet packet. However, currently the acknowledgment packet does not influence any state transitions (as we introduce in the rest of the document).

#### Sent by the baby blockchain

- UnbondingOver(seqNums): Packet sent by the baby blockchain to signalize that the unbonding period has elapsed at the baby blockchain for every validator that is demanded in the *ChangeValidatorSet* packets with sequence numbers less than or equal to *seqNum*.

  *Remark:* There exists the acknowledgment packet for the UnbondingOver packet. However, currently the acknowledgment packet does not influence any state transitions (as we introduce in the rest of the document).

## Transitions

In this section, we informally discuss the state transitions that occur in our protocol.

> We observe state transitions that are driven by an user, driven by the relayer and driven by elapsed time.

Reviewer: ```suggestion
We observe state transitions that are driven by a user (i.e., the new staking module on the parent chain), driven by the relayer and driven by elapsed time.
```

First, the staking module of the parent blockchain issues this demand by initiating the sending of the packet to the baby blockchain.
Once this packet is received by the baby blockchain, the validator set of the baby blockchain is updated to reflect the received packet and, from this point on, the "new" validator set is operating the baby blockchain.
However, once the validator set of the baby blockchain is updated, the "old" validator set starts unbonding.
Once the unbonding period for the validator set elapses, the *UnbondingOver* packet is sent to the parent blockchain.
Lastly, when the *UnbondingOver* packet is received by the parent blockchain, the stake of the validators is unfrozen.

We now present the state machine for this case.
![image](../images/data_flow.png)

## Function Definitions

### Functions provided by the staking module

In this subsection we explain which functions we assume are provided by the staking module.

```golang
// parent blockchain; used to unfreeze stake of validators associated with the change validator set demands (valSet, seqNum')
// where seqNum' <= seqNum
func unfreezeStake(seqNum)
```

> 

Reviewer: ```suggestion

- Implementation remark
    - This callback is used to signal the user (the staking module) that it can unfreeze stake because the unbonding period expired on the baby chain. It is in the responsibility of the staking module to check how much should be unfrozen depending on what evidence was received (by other means than this protocol) before.
```

func startUnbonding(seqNum, time)
```

- Expected precondition
  - The validator set of the baby blockchain reflects a change validator set demand *(V, SN)*, where *SN > seqNum*
  - End-Block method is executed for the block with bft time equals to *time*
- Expected postcondition
  - The staking module notes that the unbonding for the change validator set demand *(valSet, seqNum)* started at bft time *time*
- Error condition
  - If the precondition is violated

```golang
// baby blockchain; returns all change validator set demands that started unbonding at time T', where *time* - T' >= unbonding period
// and removes them from the data structure taking note of who is currently unbonding
// (i.e., subsequent finishUnbonding calls do not return this demand)
func finishUnbonding(time)
```

- Expected precondition
  - End-Block method is executed for the block with bft time equals to *time*
- Expected postcondition
  - Returns (and removes from the data structure) all validator set change demands *(V, SN)* such that *startUnbonding(SN, time')* is invoked, where *time - time' >= unbonding period*
- Error condition
  - If the precondition is violated

> We assume that the staking module of the baby chain takes care of a queue *validatorSetChangeQueue* which is used to store observed change validator set demands.

Reviewer: ```suggestion
**Remark.** The staking module of the baby chain needs to take care of a queue *validatorSetChangeQueue* which is used to store observed change validator set demands.
```
